File: packages/ts/core/test/fixtures/fake-provider.ts
import type { ZodType } from "zod";

import type { ProviderConfig } from "../../src/config.js";
import type { LLMProvider, StreamChunk } from "../../src/interface.js";
import type { ChatRequest, ToolResult, UnifiedResponse } from "../../src/models/index.js";

/**
 * Test double implementing `LLMProvider`. Records the `ProviderConfig` it was
 * constructed with and returns deterministic responses keyed by `providerId`,
 * so factory + orchestrator behavior is provable with zero real adapters.
 */
export class FakeProvider implements LLMProvider {
  readonly providerId: string;
  readonly config: ProviderConfig;

  constructor(config: ProviderConfig) {
    this.config = config;
    this.providerId = config.provider;
  }

  async chat(_req: ChatRequest): Promise<UnifiedResponse> {
    return {
      text: `chat:${this.providerId}`,
      usage: { promptTokens: 1, completionTokens: 1 },
      toolCalls: [],
      finishReason: "stop",
    };
  }

  async *stream(_req: ChatRequest): AsyncGenerator<StreamChunk, void, unknown> {
    yield { delta: `chat:${this.providerId}`, finishReason: "stop" };
  }

  async generateStructured<T>(_req: ChatRequest, schema: ZodType<T>): Promise<T> {
    return schema.parse({});
  }

  async invokeTool(_req: ChatRequest): Promise<UnifiedResponse> {
    return {
      text: "",
      usage: { promptTokens: 1, completionTokens: 1 },
      toolCalls: [{ id: "call_1", name: "echo", arguments: {} }],
      finishReason: "tool_use",
    };
  }

  async resumeTool(_req: ChatRequest, _results: ToolResult[]): Promise<UnifiedResponse> {
    return {
      text: `resume:${this.providerId}`,
      usage: { promptTokens: 1, completionTokens: 1 },
      toolCalls: [],
      finishReason: "stop",
    };
  }
}

File: packages/py/llmorch-core/tests/fixtures/fake_provider.py
"""Shared factory/orchestrator test double — twin of test/fixtures/fake-provider.ts."""

from collections.abc import AsyncIterator
from typing import TypeVar

from llmorch_core.config import ProviderConfig
from llmorch_core.interface import LLMProvider, StreamChunk
from llmorch_core.models import (
    ChatRequest,
    TokenUsage,
    ToolInvocationRequest,
    ToolResult,
    UnifiedResponse,
)
from pydantic import BaseModel

M = TypeVar("M", bound=BaseModel)


class FakeProvider(LLMProvider):
    """Records the ProviderConfig; returns deterministic responses keyed by id."""

    def __init__(self, config: ProviderConfig) -> None:
        self.config = config
        self.provider_id = config.provider

    async def chat(self, req: ChatRequest) -> UnifiedResponse:
        return UnifiedResponse(
            text=f"chat:{self.provider_id}",
            usage=TokenUsage(prompt_tokens=1, completion_tokens=1),
            tool_calls=[],
            finish_reason="stop",
        )

    async def stream(self, req: ChatRequest) -> AsyncIterator[StreamChunk]:
        yield StreamChunk(delta=f"chat:{self.provider_id}", finish_reason="stop")

    async def generate_structured(self, req: ChatRequest, schema: type[M]) -> M:
        return schema.model_validate({})

    async def invoke_tool(self, req: ChatRequest) -> UnifiedResponse:
        return UnifiedResponse(
            text="",
            usage=TokenUsage(prompt_tokens=1, completion_tokens=1),
            tool_calls=[ToolInvocationRequest(id="call_1", name="echo", arguments={})],
            finish_reason="tool_use",
        )

    async def resume_tool(self, req: ChatRequest, results: list[ToolResult]) -> UnifiedResponse:
        return UnifiedResponse(
            text=f"resume:{self.provider_id}",
            usage=TokenUsage(prompt_tokens=1, completion_tokens=1),
            tool_calls=[],
            finish_reason="stop",
        )
